feat: add `metadata_filters` to `filter_by` method

alvarobartt opened the pull request:
# Description

This PR adds the `pydantic.BaseModel` schemas for the `MetadataFilters` i.e. `TermsMetadataFilter`, `IntegerMetadataFilter`, and `FloatMetadataFilter` and also adds the arg `metadata_filters` to the `filter_by` method in the `RemoteFeedbackDataset` to be able to filter based on the pre-defined conditions for those metadata properties defined in the `FeedbackDataset` in Argilla.

Closes #3835

**Type of change**

- [X] New feature (non-breaking change which adds functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

- [x] Add unit tests for `TermsMetadataFilter`, `IntegerMetadataFilter`, and `FloatMetadataFilter`
- [ ] Add integration tests for the `filter_by` method with arg `metadata_filters` -> On hold because Elastic Search indexing is not working fine when triggering the tests
- [ ] Add integration tests for the `get_records` function in the SDK with the arg `metadata_filters` -> On hold because Elastic Search indexing is not working fine when triggering the tests

**Checklist**

- [ ] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [X] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

alvarobartt committed:
feat: add `{Int,Float,Terms}MetadataFilter`

alvarobartt committed:
feat: add `metadata_filters` in `filter_by`

alvarobartt committed:
feat: add `metadata_filters` to `get_records`

alvarobartt committed:
fix: remove non-existing `visible_for_annotators`

alvarobartt committed:
fix: invert `NumericMetadata{Property,Filter}` inheritance

alvarobartt committed:
Merge branch 'feature/support-for-metadata-filtering-and-sorting' of github.com:argilla-io/argilla into feat/filter-by-metadata

alvarobartt committed:
fix: `ArgillaRecordsMixin` to ignore slicing when `__len__` not implemented

alvarobartt committed:
fix: `metadata_filters` parsing in `get_records`

frascuchon commented on src/argilla/client/feedback/schemas/metadata.py:
```diff
@@ -148,16 +156,66 @@ class FloatMetadataProperty(MetadataPropertySchema):
+
+    @property
+    def server_settings(self) -> Dict[str, Any]:
+        return {"values": self.values}
```
The server expects a comma-separated list of values.

frascuchon commented on src/argilla/client/sdk/v1/datasets/api.py:
```diff
@@ -192,6 +194,9 @@ def get_records(
         params["response_status"] = response_status
 
+    if metadata_filters is not None:
+        params["metadata"] = [str(metadata_filter).replace("'", '"') for metadata_filter in metadata_filters]
```
Why do we need to `.replace` the filter content?

alvarobartt replied:
Because the single-quotes are not accepted and the API complains, while if parsing those to double-quotes everything works as expected

alvarobartt committed:
fix: add `_NumericMetadata{Property,Filter}Schema`

alvarobartt committed:
fix: `TermsMetadataFilter.server_settings` to return `str`

frascuchon commented on src/argilla/client/feedback/schemas/metadata.py:
```diff
@@ -148,16 +156,66 @@ class FloatMetadataProperty(MetadataPropertySchema):
+        exclude = {"type"}
+
+    @abstractproperty
+    def server_settings(self) -> Dict[str, Any]:
```
I understand the `server_settings` purpose, but since the filter schema has no settings and they will be used as part of the query param in the URL, maybe we should change the signature:
```python
def as_server_repr(self) -> str:
# or
def to_server_schema(self) -> str:

```

alvarobartt replied:
True, I'll change it now for the `MetadataFilters`, thanks!

frascuchon commented on src/argilla/client/feedback/schemas/metadata.py:
```diff
@@ -148,16 +156,66 @@ class FloatMetadataProperty(MetadataPropertySchema):
+        exclude = {"type"}
+
+    @abstractproperty
+    def server_settings(self) -> Dict[str, Any]:
```
```suggestion
    def server_settings(self) -> str:
```

alvarobartt committed:
fix: handle no-records during `_fetch_records`

alvarobartt committed:
refactor: `server_settings` to `query_string`

alvarobartt committed:
fix: mandatory args in `validate_numeric_metadata_property_bounds`

alvarobartt committed:
(docs,fix): add docstrings and fix type-hints

alvarobartt committed:
fix: ensure at least one is provided in `validate_numeric_metadata_filter_bounds`

alvarobartt committed:
fix(revert): ensure `min` and `max` is not None

alvarobartt committed:
test: add `{Terms,Integer,Float}MetadataFilter` unit tests

alvarobartt committed:
fix: remove `RemoteFloatMetadataProperty.visible_for_annotators`

alvarobartt committed:
style(test): improve `tests.factories` usage

alvarobartt committed:
fix(style): add missing line break

alvarobartt committed:
test: add `test_get_records_using_metadata_filters`

alvarobartt committed:
fix(test): add `owner` role to `pytest.mark.parametrize`

alvarobartt committed:
test: add notes on failing test for `metadata_filters`

frascuchon commented on src/argilla/client/feedback/dataset/remote/mixins.py:
```diff
@@ -84,6 +95,10 @@ def __getitem__(
                 ]
             )
+        if len(records) == 0:
+            raise RuntimeError(
```
`IndexError` instead?

frascuchon commented on src/argilla/client/feedback/schemas/validators.py:
```diff
@@ -45,3 +45,17 @@ def validate_numeric_metadata_property_bounds(cls: "BaseModel", values: Dict[str
+    if le is None and ge is None:
+        raise ValueError(f"`{cls.__name__}` with name={name} must have at least one of `le` or `ge`")
+
+    if le is not None and ge is not None and le >= ge:
```
If not, we cannot support filtering records with just one value in a numeric metadata property. For example: fetching records with `metadata.value == 10`. We can define in terms of filter as : `le=10, ge=10`
```suggestion
    if le is not None and ge is not None and le > ge:
```

frascuchon commented on tests/unit/client/feedback/schemas/test_metadata.py:
```diff
@@ -251,3 +254,124 @@ def test_float_metadata_property_errors(
+            "1 validation error for IntegerMetadataFilter\n__root__\n  `IntegerMetadataFilter` with name=integer-metadata-filter must have at least one of `le` or `ge`",
+        ),
+        (
+            {"name": "int-metadata-filter", "le": 5, "ge": 5},
```
This example must be accepted until we have another way o support equal conditions for this propertites

frascuchon commented on tests/unit/client/feedback/schemas/test_metadata.py:
```diff
@@ -251,3 +254,124 @@ def test_float_metadata_property_errors(
+            "1 validation error for FloatMetadataFilter\n__root__\n  `FloatMetadataFilter` with name=float-metadata-filter must have at least one of `le` or `ge`",
+        ),
+        (
+            {"name": "float-metadata-filter", "le": 5.0, "ge": 5.0},
```
Same here. Also, since you're defining float values, I suggest giving some decimal value instead of 0. This could help find unexpected round errors.

alvarobartt committed:
fix(test): remove non-working tests

alvarobartt committed:
fix: validator for `le` and `ge` to be `le > ge`

alvarobartt committed:
refactor: change `RuntimeError` to `IndexError`

alvarobartt committed:
fix(test): update `test_metadata.py`

alvarobartt committed:
docs: update `CHANGELOG.md`

alvarobartt committed:
feat: add `{Terms,Integer,Float}MetadataFilter` imports to `__init__`

alvarobartt committed:
fix: add missing `metadata_properties` in `pull`

gabrielmbmb commented on src/argilla/client/feedback/schemas/validators.py:
```diff
@@ -42,6 +42,20 @@ def validate_numeric_metadata_property_bounds(cls: "BaseModel", values: Dict[str
+        raise ValueError(f"`{cls.__name__}` with name={name} must have at least one of `le` or `ge`")
+
+    if le is not None and ge is not None and le > ge:
+        raise ValueError(f"`{cls.__name__}` with name={name} cannot have `ge` less than `le`")
```
```suggestion
    if le is not None and ge is not None and le < ge:
        raise ValueError(f"`{cls.__name__}` with name={name} cannot have `le` less than `ge`")
```

alvarobartt committed:
fix: reverse `le` with `ge` in `validate_numeric_metadata_filter_bounds`

Co-authored-by: Gabriel Martín Blázquez <gmartinbdev@gmail.com>

frascuchon committed:
Refactor: several improvements and bug fixes (#3847)

# Description

This PR covers some code refactoring and bug fixes discovered during the
Python client testing.

* Allow creation records with optional fields not provided or provided
as `None` value.
* Move the client filters setup to the `FilteredRemoteFeedbackRecords`
class. Also, the `filters` have been flattening since the class needs to
work with `response_status` and `metadata_filters` separately.
* Redefine some `ArgillaMixing` methods as `@staticmethod` since no
internal status was needed
* Type-hint and methods signature review.

**Type of change**

- [X] Refactor (change restructuring the codebase without changing
functionality)
- [X] Improvement (change adding some improvement to an existing
functionality)

**Checklist**

- [ ] I added relevant documentation
- [x] I followed the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
